src/main/java/com/commercetools/sync/customobjects/CustomObjectSync.java

```diff
@@ -0,0 +1,329 @@
+package com.commercetools.sync.customobjects;
+
+import com.commercetools.sync.commons.BaseSync;
+import com.commercetools.sync.commons.exceptions.SyncException;
+import com.commercetools.sync.customobjects.helpers.CustomObjectCompositeIdentifier;
+import com.commercetools.sync.customobjects.helpers.CustomObjectSyncStatistics;
+import com.commercetools.sync.customobjects.utils.CustomObjectSyncUtils;
+import com.commercetools.sync.services.CustomObjectService;
+import com.fasterxml.jackson.databind.JsonNode;
+import io.sphere.sdk.customobjects.CustomObject;
+import io.sphere.sdk.customobjects.CustomObjectDraft;
+import org.apache.commons.lang3.tuple.ImmutablePair;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+
+import static com.commercetools.sync.commons.utils.SyncUtils.batchElements;
+import static java.lang.String.format;
+import static java.util.Optional.ofNullable;
+import static java.util.concurrent.CompletableFuture.completedFuture;
+import static java.util.function.Function.identity;
+import static java.util.stream.Collectors.toMap;
+import static java.util.stream.Collectors.toSet;
+
+/**
+ * This class syncs custom object drafts with the corresponding custom objects in the CTP project.
+ */
+public class CustomObjectSync extends BaseSync<CustomObjectDraft<JsonNode>,
+    CustomObjectSyncStatistics, CustomObjectSyncOptions> {
+
+    private static final String CTP_CUSTOM_OBJECT_FETCH_FAILED =
+        "Failed to fetch existing customObjects with keys: '%s'.";
+    private static final String CTP_CUSTOM_OBJECT_UPSERT_FAILED =
+        "Failed to create/update customObjects with key: '%s'. Reason: %s";
```

Comment:
imo it should be an error message for singular, not for custom objects..

Reply:
https://github.com/commercetools/commercetools-sync-java/pull/569/commits/893a6d19adde88df3b3d5a973800c0bf0cb5dadc